v1/storage/path.go

```diff
@@ -133,11 +133,22 @@ func (p Path) Ref(head *ast.Term) (ref ast.Ref) {
 }
 
 func (p Path) String() string {
-	buf := make([]string, len(p))
-	for i := range buf {
-		buf[i] = url.PathEscape(p[i])
+	if len(p) == 0 {
+		return "/"
 	}
-	return "/" + strings.Join(buf, "/")
+
+	l := 0
+	for i := range p {
+		l += len(p[i]) + 1
+	}
+
+	sb := strings.Builder{}
+	sb.Grow(l)
+	for i := range p {
+		sb.WriteByte('/')
+		sb.WriteString(url.PathEscape(p[i]))
```

Comment:
If this causes escaping, I suppose the only ill effect is that the above `sb.Grow(l)` wasn't enough and `sb` needs to grow further?

Since you're interested in performance in this PR, I assume you've confirmed this is better or equal in performance compared to the old method for the general/common case. How about the worst case, where `sb` needs to grow further (maybe more than once)?